qutip/core/cy/coefficient.pyx

```diff
@@ -365,167 +373,180 @@ def coeff(t, args):
     """
-    A :obj:`Coefficient` built from a cubic spline interpolation of a numpy array.
+    A :obj:`Coefficient` built from an interpolation of a numpy array.
 
     Parameters
     ----------
     coeff_arr : np.ndarray
         The array of coefficient values to interpolate.
 
     tlist : np.ndarray
-        An array of times corresponding to each coefficient value. The times must be
-        increasing, but do not need to be uniformly spaced.
+        An array of times corresponding to each coefficient value. The times
+        must be increasing, but do not need to be uniformly spaced.
+
+    order : int
+        Order of the interpolation. ``0`` use previous or left value.
     """
-    cdef int n_t, constant
+    cdef int order
     cdef double dt
     cdef double[::1] tlist
-    cdef complex[::1] coeff_arr, second_derr
-    cdef object tlist_np, coeff_np, second_np
-
-    def __init__(self, coeff_arr, tlist, _second=None, _constant=None):
-        self.tlist_np = tlist
-        self.tlist = tlist
-        self.coeff_np = coeff_arr
-        self.coeff_arr = coeff_arr
-        if _second is None:
-            self.second_np, self.constant = _prep_cubic_spline(coeff_arr,
-                                                                tlist)
+    cdef complex[:, :] poly
+    cdef object np_arrays
+
+    def __init__(self, coeff_arr, tlist, int order):
+        tlist = np.array(tlist, dtype=np.float64)
+        coeff_arr = coeff_arr.astype(np.complex128)
+
+        if coeff_arr.ndim != 1:
+            raise ValueError("The array to interpolate must be a 1D array")
+        if coeff_arr.shape != tlist.shape:
+            raise ValueError("tlist must be the same len "
+                             "as the array to interpolate")
+        if order < 0:
+            raise ValueError("order must be a positive integer")
+
+        order = min(order, len(tlist) - 1)
+
+        if order == 0:
+            self.np_arrays = (tlist, coeff_arr.reshape((1, -1)))
+        elif order == 1:
+            self.np_arrays = (
+                tlist,
+                np.vstack([
+                    np.diff(coeff_arr, append=-1) / np.diff(tlist, append=-1),
+                    coeff_arr
+                ]))
+        elif order >= 2:
+            # Use scipy to compute the spline and transform it to polynomes
+            # as used in scipy's PPoly which is easier for us to use.
+            spline = make_interp_spline(tlist, coeff_arr, k=order)
+            # Scipy can move knots, we add them to tlist
+            ts = np.sort(np.unique(np.concatenate([spline.t, tlist])))
+            a = np.arange(spline.k+1)
+            a[0] = 1
+            fact = np.cumprod(a)
+            poly = np.concatenate([
+                spline(ts, i) / fact[i]
+                for i in range(spline.k, -1, -1)
+            ]).reshape((spline.k+1, -1))
+            self.np_arrays = (ts, poly)
+
+        self._prepare()
+
+    def _prepare(self, dt=None):
```

Comment:
I think we should pass `np_arrays` to `_prepare` and then have `_prepare` set all of the attributes rather than having most of them set here and having all the other functions need to remember to set `np_arrays`.

I would also prefer `_prepare(np_tlist, np_poly, dt)` so that it's clear that nd_arrays should have exactly two elements (although this also means changing `__reduce__` and `restore` to match).